### DISCUSSION TITLE
Some additional features, mostly for modding

### DISCUSSION THREAD

**Discussion by 4164696f73** [Try/OpenGothic#226] [Ideas]
There's a few things "wrong" with the original Gothic design that, judging by what modders do, people dislike. I'd like everyone to send in here what they hate/dislike and should be an option to change.

Disclaimer: This is mostly for modding purposes, though if the author of the project chooses to add those as an option in the settings, it's also welcome. I didn't really think of many other things because I haven't been playing Gothic much recently. Also, if people are going to request things, do remember to only mention things that are tedious to mod (aka take a long time) or require way too much hacking to achieve. And make those QoL requests, not game changing ones.

1. Printsscreens overlapping in certain cases:
    In plenty of cases, if you gain experience, items (via dialogue), and probably other stuff, the printscreens are printed one on top of each other, making it impossible to know what just happened, what we got etc. Modders, I'd dare to say "always", replace the original print with their own, which typically is placed on the left side of the screen and moves the last print(s) to the top of the screen to show the new message, effectively creating lines of text on the screen with all those information, instead putting them one on top of each other and technically giving you no information about anything because all the prints have the same X and Y coordinates.
    - Simply add another type of PrintScreen (for example called PrintScreenA - A for "animation") that takes all the same parameters as regular PrintScreen (for purposes of mass editing in the mod, where you'd just swap PrintScreen(...) to PrintScreenA(...) and not have to bother with changing every file that contains the PrintScreen. The animated behavior would be similar to Print().
    - Add a modifier variable to PrintScreen that modders can change in the mod files. For example the variable would be a boolean and called "PrintScreenAnimation", which by default is set to 0, which means all prints behave as they do in unmodded, clean Gothic. If set to 1 (or true), all PrintScreens are now printed on the left side on the screen, and animated to move the oldest messages to the top of the screen when a new message is added. The animated behavior would be similar to Print(). Can add more customisation, for example a variable to use new coordinates (global change, change one variable [or two] to move all the prints wherever you want, globally) etc.
2. AI modification:
    We all know that some monster's fighting styles are weird, or stupid, or we just want to add a new monster with new behavior, and we don't have any way to change the AI. Sure, we can hack with some third party tools like Ikarus etc, though it's a ton of work and better to just use an existing AI at this point.
    - Haven't really gotten a good idea of how to implement it. My initial thought was to add a new variable to npc class, which would contain a function of the new AI/FAI (though regular AI, for the most part, can be done with Daedalus), and then fill the FAI_tactics (or whatever was the variable name) with the new identifier (e.g. FAI_CACTUS [don't ask]). There probably are other ways to do it better.

I'm actually really interested in OpenGothic and am thinking about using it if I'd ever get back to modding, though I'd like to avoid modifying stuff engine-wise, hence I made this post (so that everyone can get the engine themselves and just use the mods without the need of redownloading the engine).

**Comment by Try** [comment]
Hi, @4164696f73  !

That's is a nice idea to expand script api. Also would be nice to establish some general view:
* do we want to keep compatibility with vanilla? (Thru Ikarus/Union) 
* establish overall view of api would be nice - better then soup of different suggestions 

> Printsscreens overlapping in certain cases:

Yep, we can also take a step forward and do a proper UI library, with ListView/Canvas/Timer

> AI modification ...  add a new variable to npc class

There is a `fight_tactic` already in gothic. It work like that:
1. Script invokes AI_Atack
2. In AI_Atack routine engine will access instance with name `FA_%s_%d`, where %d - is integer value of `fight_tactic`, and %s - combat situation
3. In, for exapmple, `C_FightAI` there is only one field `move[6]` - engine will pick random defined move and perform
4. In OpenGothic, there is a extra small queue, that will be filled by `move[rand]`. For example `MOVE_TRIPLEATTACK` -> 3 queue actions

In principle we can introduce api, like `AI_Atack_v2(int explicit_move)` to enable explicit action from script.

**Reply by 4164696f73** [reply]
I'm just providing my opinions here, some people might disagree. I fee like putting it in the first line because some people might not read till the end. 
Also, in the end it's up to the project maintainer(s), so we don't decide what the project should look like! We just give opinions and our views on modding! :) 


> * do we want to keep compatibility with vanilla? (Thru Ikarus/Union)

That would be good to have, though you could (if you don't mind spending additional tons of times on it) implement hooking from Daedalus level just like Ikarus does - just hooking in general, so that you still could do what Ikarus does, without having it (though I'd assume it would require someone to rewrite the hole Ikarus package).

Though I'd personally not want to use Ikarus itself while modding, but imitate its behavior and how it works. Ikarus is rather slow with some of its loops and hooks, so if the engine itself offered hooking by default, everything would be sped up. Not sure how you'd be able to implement loops, which actually are the slowest out of the whole Ikarus package (from my experience from back from 2015/2016 - might be way better now, I don't know).

Can't talk about Union as I never used it. I believe it's the "extension" that allows you to write custom libraries? If so, support with that would be insanely good to have.

I was also thinking about allowing to modify the engine from Daedalus level, like, you'd do something similar to "engine.DamageFunc.damage = (str*damage)/prot" and now the damage would always equal to that, so that we no longer need Ikarus, or "engine.Print = MyNewPrint()", but I doubt you'd wanna add it to the base game as it would take a lot of time and do what external tools already do. 
Basically, you'd be able to hook the engine functions and whatnot from Daedalus level but modify it using just the github project as a reference instead of Assembly. But I do believe this is a stupid idea as it would take way too long to implement. Also, kind of hard to do, I'd assume, so I doubt you'd even think of it :) 

In general, I do believe that compatibility with at least Union would be "mandatory" (with a grain of salt), because (if that's the project I'm thinking of) some people actually spent a lot of time rewriting everything they had in Ikarus to Union, and it would be a waste. 
Also, if that'd be possible for you, you could just make OpenGothic read custom libraries, by default, from a certain directory. For example, someone has some modifications via libraries, so they just install OpenGothic, create a new directory (let's say in System) called "ModLibraries" and put everything they had there, and now every time they launch the game, those libraries are read. Also mod compatibility would be great, where if something is in "System/ModLibraries", the game reads them all despite of mods, but if you have "System/ModLibraries/Returning", only mods with the name "Returning" would make the game read those libraries. Aka no more Union needed to attach the libraries, or whatever it does with them. :P 
Ikarus compatibility is fine, but I don't think people would be sad if they couldn't use Ikarus and had to rewrite their functions for OpenGothic, if it provided a faster functionality.

> Yep, we can also take a step forward and do a proper UI library, with ListView/Canvas/Timer
 
Yeah that would be really amazing to have!

> 
> There is a `fight_tactic` already in gothic. It work like that:
> 
> 1. Script invokes AI_Atack
> 2. In AI_Atack routine engine will access instance with name `FA_%s_%d`, where %d - is integer value of `fight_tactic`, and %s - combat situation
> 3. In, for exapmple, `C_FightAI` there is only one field `move[6]` - engine will pick random defined move and perform

From what I remember, I never could make it via Daedalus level, but I might've not tried hard enough. Also, plenty of modders always said that you can not modify any AI in any way, aside of the routines, which is why I never really spent much time on it. 

Also, I remember not being able to write conditions where the enemy responds to certain hits and moves, which is mostly what I was talking about (probably wasn't clear enough, sorry). For example "if human does combo attack on wolf -> loop <dodge backwards> x3". But again, I probably was discouraged to try hard enough when I heard people saying it's not possible.

*Secretly installs Gothic to play with that*

> 4. In OpenGothic, there is a extra small queue, that will be filled by `move[rand]`. For example `MOVE_TRIPLEATTACK` -> 3 queue actions

If that means you'd be able to make enemies use certain combos with weapons, that means the game will be more enjoyable. I like that!

> In principle we can introduce api, like `AI_Atack_v2(int explicit_move)` to enable explicit action from script.

That would be nice to have, honestly! You could make some fancy things with it, like an enemy that blocks more attacks the lower their HP is, responds to certain actions done by their enemy (for example hook the AttackForward function and make an npc respond with perma blocking to it etc) and many other fancy things.


The more OpenGothic expands on modding the better, honestly. I really do believe that your project is the future of modding if we, modders, contribute with ideas like I had. Obviously there are other things modders would want to have easier (like modifying the engine whenever they want without hacking with Assembly etc). 
Downloading your project and modifying stuff as we please is another way to mod the game, but there are issues with it where some mods will have different engine modifications and hence you'd need multiple .exe on your PC and launch them depending on which mod you wanna play, and people can forget and launch the wrong .exe or whatever, which is why it'd rather be better if we could modify it from Daedalus level, so that only one mod package is changed, aka one file, instead of the mod and .exe (or potentially more files).
